cmd/uplink/cmd/cp_test.go

```diff
@@ -0,0 +1,125 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package cmd_test
+
+import (
+	"crypto/rand"
+	"log"
+	"os"
+	"strconv"
+	"testing"
+
+	"storj.io/storj/internal/memory"
+	"storj.io/storj/internal/s3client"
+)
+
+const (
+	bucket                = "testbucket"
+	uplinkEncryptionKey   = "supersecretkey"
+	storjSimSatelliteAddr = "127.0.0.1:10000"
+)
+
+func setup() s3client.Client {
+	var conf s3client.Config
+	conf.Satellite = storjSimSatelliteAddr
+	conf.EncryptionKey = uplinkEncryptionKey
+	conf.APIKey = os.Getenv("storjSimApiKey")
+
+	client, err := s3client.NewUplink(conf)
+	if err != nil {
+		log.Fatalf("failed to create NewUplink: %+v\n", err)
+	}
+	err = client.MakeBucket(bucket, "")
+	if err != nil {
+		log.Fatalf("failed to create bucket %q: %+v\n", bucket, err)
+	}
+	return client
+}
+
+func teardown(client s3client.Client) {
+	for _, bm := range benchmarkCases {
+		deleteUploadedFiles(bm.name)
+	}
+
+	err := client.RemoveBucket(bucket)
+	if err != nil {
+		log.Fatalf("failed to remove bucket %q: %+v\n", bucket, err)
+	}
+}
+
+func deleteUploadedFiles(name string) {
+	for filePath := range uploadedFiles[name] {
+		err := client.Delete(bucket, filePath)
+		if err != nil {
+			log.Printf("failed to delete file %q: %+v\n", filePath, err)
+		}
+	}
+}
+
+var client = setup()
+
+var benchmarkCases = []struct {
+	name     string
+	filesize memory.Size
+}{
+	{"100B", 100 * memory.B},
+	{"1MB", 1 * memory.MiB},
+	{"10MB", 10 * memory.MiB},
+	{"100MB", 100 * memory.MiB},
+	{"1G", 1 * memory.GiB},
+}
+
+var uploadedFiles = map[string]map[string]struct{}{}
+
+func BenchmarkUpload(b *testing.B) {
+	for _, bm := range benchmarkCases {
+		b.Run(bm.name, func(b *testing.B) {
+			file := createRandomBytes(bm.filesize)
```

Comment:
or, just making the random bytes once at the top?

Reply:
i think i meant to use math/rand, darn automatic imports. lets me test around with whats faster. 

i think the intended end user is jenkins, so not windows

Reply:
It might be possible to reuse the functionality in https://github.com/storj/storj/blob/b0ca12e31eef8922a68b79ff1ffe46a5d637255e/scripts/test-aio.sh#L50-L54 (and related scripts). You'd have to move those parts into a shell script, though.

Reply:
I copied that bash script and ran it locally:

```
# testfile.sh
#!/bin/sh
TMP_DIR=$(mktemp -d -t tmp.XXXXXXXXXX)
mkdir -p "$TMP_DIR"

echo "temp dir:"
echo "$TMP_DIR"

random_bytes_file(){
    size=$1
    output=$2
    dd if=/dev/urandom of="$output" count=1 bs="$size" >/dev/null 2>&1
}
echo "=> Making test files"
random_bytes_file      1x1024x1024 "$TMP_DIR/one-mb"
random_bytes_file      10x1024x1024 "$TMP_DIR/ten-mb"
random_bytes_file      100x1024x1024 "$TMP_DIR/one-hundred-mb"
random_bytes_file      1000x1024x1024 "$TMP_DIR/one-gb"

echo
echo "cleanup"
rm -rf "$TMP_DIR"
````


```
$ time ./testfile.sh
temp dir:
/var/folders/s_/n4czty1x4tj297495m66bcsc0000gp/T/tmp.XXXXXXXXXX.UZIxhRGj

=> Making test files

cleanup

real	0m29.592s
user	0m0.009s
sys	0m29.190s
```

Then added  this function to the cp_test.go code and ran locally:
```
var benchmarkCases = []struct {
    name     string
    filesize memory.Size
}{
    {"100B", 100 * memory.B},
    {"1MB", 1 * memory.MiB},
    {"10MB", 10 * memory.MiB},
    {"100MB", 100 * memory.MiB},
    {"1G", 1 * memory.GiB},
}

var files = createFiles()

func createFiles() map[string][]byte {
    start := time.Now()
    files := make(map[string][]byte)
    for _, x := range benchmarkCases {
        file := createRandomBytes(x.filesize)
        files[x.name] = file
    }
    elapsed := time.Since(start)
    log.Printf("createFiles took %s", elapsed)
    return files
}
```

```
$ go test -bench . -benchmem ./cmd/uplink/cmd/

	2019/05/23 13:14:13 createFiles took 1.834741827s

[…]
```

Looks like `dd if=/dev/urandom` took ~30s versus `math/rand` taking 1.8s

Reply:
DATA-BASED DECISIONS ARE THE BEST DECISIONS! Woohoo.

I figured math/rand was faster, but wow, that's impressive.

Reply:
yeah i was surprised too!